fix(docs): Update the activation model (#4112)
tempoxyz/tempo: tips/tip-1063.md

## Abstract

> This TIP defines feature flags for Tempo protocol changes, decoupling code release from onchain activation.

Author: ```suggestion
This TIP defines feature flags for Tempo protocol changes, decoupling individual feature activation from hardfork bundles.
```

Feature flags separate code distribution from behavior activation. Validators can upgrade over time, and each feature can activate only after enough of the active validator set is running a supporting Tempo version. This means releases can continue on a regular cadence, while activation can be scheduled, delayed, or cancelled independently.

# Specification

## 1. Scope

This TIP defines the feature flag mechanism for Tempo protocol features.

A feature flag controls when a specific protocol change can become active. The code can ship inactive first, then become active after it has been scheduled by the multisig and enough active validators have reported a supporting Tempo version.

A protocol feature is appropriate when the change affects consensus validation, state transition behavior, transaction validity, system contract behavior, precompile behavior, block building, RPC simulation, or other protocol-facing logic.

This TIP makes feature flags the default path for protocol changes that can be shipped before activation.

> This TIP does not define the behavior of any individual future feature. Each feature still needs its own TIP or approved feature record describing the behavior being activated, operator impact, tests, and rollout requirements.

Author: ```suggestion
This TIP does not define the behavior of any individual future feature. Each feature still needs its own TIP or approved feature record describing the behavior being activated, compatibility impact, tests, and rollout requirements.
```

## 2. Protocol Features

A protocol feature is a specific protocol behavior that can be tracked and activated independently.

Protocol features are listed in a source-controlled feature manifest. The manifest is a single file that contains the feature metadata used by release tooling and the feature registry.

Each feature entry includes:

- `feature_id`: a globally unique, lowercase, dot-separated feature identifier
- `description`: a short description of the behavior
- `tip`: the TIP that specifies the behavior
- `minimum_supported_version`: the first Tempo release expected to support the feature, for operator and tooling visibility

> - `status`: the feature's source-controlled review status

Author: ```suggestion
- `status`: the feature's manifest status, such as draft, approved, deprecated, or cancelled
```

> - `minimum_supported_version`: the first Tempo release expected to support the feature

Author: ```suggestion
- `minimum_supported_version`: the first Tempo release expected to support the feature, for operator and tooling visibility
```

Feature IDs are append-only. Once a `feature_id` has been assigned to a feature, that ID is not reused for a different behavior.

The feature's `minimum_supported_version` is the first Tempo release expected to validate, build, simulate, and serve protocol behavior correctly after that feature becomes active.

This model infers feature support from reported Tempo versions. It assumes release support is mostly monotonic: if an official release supports a feature, later official releases are expected to keep supporting that feature unless the feature is cancelled before activation or superseded by a later protocol feature.

## 3. Validator Version Reporting

Validators report the Tempo version they are running. The reported version is the readiness signal used for features whose `minimum_supported_version` is less than or equal to that version.

The exact reporting path is an implementation detail. The report must be deterministic from chain history before it can count toward quorum.

Quorum is calculated from reported validator versions.

> The latest reported version replaces the validator's previous reported version. If a validator reports a newer version, support counts for features whose minimum version is now satisfied must be updated. If a validator reports a version below a feature's minimum supported version, that validator no longer contributes support for that feature.

Author: ```suggestion
Each validator's latest reported version is used for quorum. A validator contributes support for a feature only when its reported version is greater than or equal to the feature's `minimum_supported_version`.
```

> Version comparison must be deterministic from chain history. The implementation can encode Tempo releases as monotonically increasing release IDs or use a registry of known release IDs, but it must not depend on ambiguous local string parsing.

Author: ```suggestion
Tempo versions used for quorum must have a deterministic ordering.
```

Implementations must expose enough information to determine validator versions and feature support counts.

> Validator set changes must be reflected in feature support counts. A validator entering the active set counts only after it is eligible under the validator set rules and has a reported version. A validator leaving the active set no longer contributes to quorum.

Author: ```suggestion
Only eligible active validators count toward quorum.
```

When the scheduling transaction is accepted, the registry sets:

```text
activation_timestamp = requested_activation_timestamp
```

The feature becomes active once the activation timestamp has been reached:

```text
block.timestamp >= activation_timestamp
```

Activation is applied during the pre-execution phase at the start of block processing. Before user transactions execute, protocol handlers check scheduled inactive features.

If a feature's activation timestamp has been reached, the pre-execution phase marks the feature active and emits the activation event. User transactions in that block then observe the feature as active.

Quorum is not re-evaluated at activation time. Validators that have not upgraded to a supporting release by the activation timestamp must fail closed before processing blocks where the feature is active.

`activation_timestamp` records the timestamp at which the feature becomes active.

Activation must be deterministic from chain history. Given the same chain, all nodes must compute the same active feature set for every block.

> Scheduling without quorum is outside the scope of this TIP.

Author: ```suggestion
```
